File: Cisco/59.spiral-matrix-ii.java
/**
 * Algorithm:
 * Fill the matrix using the same four-boundary spiral walk as traversal:
 * right, down, left, up, then shrink inward.
 *
 * Complexity:
 * Time O(n^2), extra space O(1) excluding output.
 */
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int top = 0;
        int bottom = n - 1;
        int left = 0;
        int right = n - 1;
        int value = 1;
        while (top <= bottom && left <= right) {
            for (int c = left; c <= right; c++) {
                matrix[top][c] = value++;
            }
            top++;
            for (int r = top; r <= bottom; r++) {
                matrix[r][right] = value++;
            }
            right--;
            for (int c = right; c >= left; c--) {
                matrix[bottom][c] = value++;
            }
            bottom--;
            for (int r = bottom; r >= top; r--) {
                matrix[r][left] = value++;
            }
            left++;
        }
        return matrix;
    }
}

File: companies/Capital One/59.spiral-matrix-ii.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        /*
        Approach:
        Fill rings from the outside inward using top, bottom, left, and right
        boundaries. Increment the value after placing it in each cell.

        Complexity: O(n^2) time and O(1) extra space beyond the output matrix.
        */
        vector<vector<int>> ans(n, vector<int>(n));
        int top = 0, bottom = n - 1, left = 0, right = n - 1, value = 1;
        while (top <= bottom && left <= right) {
            for (int c = left; c <= right; ++c) ans[top][c] = value++;
            ++top;
            for (int r = top; r <= bottom; ++r) ans[r][right] = value++;
            --right;
            for (int c = right; c >= left; --c) ans[bottom][c] = value++;
            --bottom;
            for (int r = bottom; r >= top; --r) ans[r][left] = value++;
            ++left;
        }
        return ans;
    }
};
